```{r}
library(data.table)
library(ggplot2)
library(ASRgenomics)
library(AGHmatrix)
library(rrBLUP)
library(tidyverse)
library(asreml)
setDTthreads(80)

```

```{r}
# Load full marker dataset
markers <- fread("CDBN_Filter_150Ct_MAF5per_CDBN_001_pruned_PLUS.SNPsScore.Numeric.raw")
markers <- as.data.frame(markers)
markers <- markers[, -c(1,3:6)]  # Remove unnecessary columns
colnames(markers) <- gsub("_[A-Z]$", "", colnames(markers))

# Load phenotypic data
pheno <- fread("DF.WYPO.EBLUPS.txt")
pheno$name<-as.factor(pheno$name)
n.pheno <- pheno$name

# Filter genotypes with phenotypes
markers <- markers[markers$IID %in% n.pheno,]
names <- markers$IID
markers <- markers[, -1]
rownames(markers) <- names
markers <- as.matrix(markers)
mode(markers) <- "numeric"
```

```{r}
# Load conservation scores
conservation_scores <- fread("Deleterious_weights.tsv")  # Columns: SNP, Score
conservation_scores <- conservation_scores[, c(1, 5)]

# Merge SNP conservation scores with genotype matrix
snp_info <- data.frame(SNP = colnames(markers))
snp_info <- merge(snp_info, conservation_scores, by = "SNP")

```


```{r}
# Define loop parameters
num_markers <- 7000  # Subset size
n <- nrow(markers)
nt <- n * 0.3  # 30% test set
runs <- 50

# Storage for results
corr.val <- matrix(NA, nrow = runs, ncol = 10)

for (i in 1:10) {  # Loop over 10 different marker subsets
  set.seed(i)
  selected_markers <- sample(ncol(markers), num_markers)
  M.subset <- markers[, selected_markers]
  
  # Filter markers and construct weighted G-matrix
  M_filter <- qc.filtering(M = M.subset, maf = 0.04, marker.callrate = 0, ind.callrate = 0.20, impute = FALSE)
  M.clean <- M_filter$M.clean
  
  
# Include scores as weights
  
  for (j in 1:runs) {
    set.seed(j + as.integer(Sys.time()))
    tst <- sample(1:n, size = nt, replace = FALSE)
    yNA <- pheno$DF_BLUE
    yNA[tst] <- NA
  

  n_filtered <- colnames(M.clean)
  scores <- conservation_scores[conservation_scores$SNP %in% n_filtered]
  
  match_idx <- match(n_filtered, conservation_scores$SNP)
  W <- conservation_scores$Score[match_idx]
  W[is.na(W)] <- 0
  W <- W + 0.001  # Shift effects
  
  # Compute G-matrix
  W_matrix <- Gmatrix(SNPmatrix = M.clean, method = "VanRaden", weights = W) # Depending if it is weighted GBLUP 
  Ghat.blend <- G.tuneup(G = W_matrix, blend = TRUE, pblend = 0.02)$Gb
  Ginv.sparse <- G.inverse(G = Ghat.blend, sparse = TRUE)$Ginv
  
  model <- asreml(fixed = DF_BLUE ~ 1, random = ~vm(name, source = Ginv.sparse), data = pheno[-tst,])
    
   
  preds <- predict(model, classify = "name")$pvals
  preds <- preds[preds$name %in% pheno$name[tst], ]
  merged <- merge(preds, pheno[tst, c("name", "DF_BLUE")], by = "name")
  corr.val[j, i] <- cor(merged$DF_BLUE, merged$predicted.value, use = "complete.obs")
  }
}
  
# Compute mean and standard errors
mean_corr <- colMeans(corr.val, na.rm = TRUE)
SE_corr <- apply(corr.val, 2, function(x) sd(x, na.rm = TRUE) / sqrt(runs))

# Create summary table
summary_table <- data.frame(
  Set = 1:10,
  mean_corr = mean_corr,
  SE_corr = SE_corr
)

# Print results
table_output <- capture.output(print(summary_table, row.names = FALSE))
writeLines(table_output)

View(summary_table)
mean(summary_table$mean_corr)


  
```



