
```{r}
library(data.table)
library(ggplot2)
library(ASRgenomics)
library(AGHmatrix)
library(asreml)
library(tidyverse)
setDTthreads(80)

# Load full marker dataset
markers <- fread("CDBN_Filter_150Ct_MAF5per_CDBN_001_pruned_PLUS.SNPsScore.Numeric.raw")
markers <- as.data.frame(markers)
markers <- markers[, -c(1,3:6)]  # Remove unnecessary columns

# Load phenotypic data
pheno <- fread("DF.WYPO.EBLUPS.txt")
pheno$Taxa<-as.factor(pheno$Taxa)
n.pheno <- pheno$Taxa

# Filter genotypes with phenotypes
markers <- markers[markers$IID %in% n.pheno,]
names <- markers$IID
markers <- markers[, -1]
rownames(markers) <- names
markers <- as.matrix(markers)
mode(markers) <- "numeric"

## Only for DF
pheno=pheno[pheno$Taxa %in% rownames(markers),]

# Define loop parameters
num_markers <- 7000  # Subset size
n <- nrow(markers)
nt <- n * 0.3  # 30% test set
runs <-50

# Storage for results
corr.val <- matrix(NA, nrow = runs, ncol = 10)

for (i in 1:10) {  # Loop over 10 different marker subsets
  set.seed(i)
  selected_markers <- sample(ncol(markers), num_markers)
  M.subset <- markers[, selected_markers]
  
  # Filter markers and construct weighted G-matrix
  M_filter <- qc.filtering(M = M.subset, maf = 0.04, marker.callrate = 0, ind.callrate = 0.20, impute = FALSE)
  M.clean <- M_filter$M.clean
  
  snpsWtScores<-fread("genotypes/CDBN_SNPsWithScore.txt", header = F)
  weights <- fread("CDBN_snp_effects_DF_Score_rrBLUP.txt", sep = "\t")
  weights <- weights[weights$SNP %in% snpsWtScores$V1,]
  scores=fread("CDBN_prediction_weights.tsv")
  weights=inner_join(weights, scores, by="SNP")
  weights$Effect2 <- (abs(weights$Effect)) * weights$weight
  n_filtered <- colnames(M.clean)
  n_filtered <- gsub("_[^_]+$", "", n_filtered)
      
  W <- ifelse(is.na(match(n_filtered, weights$SNP)), 0, weights$Effect2[match(n_filtered, weights$SNP)])
  W <- W + 0.001  # Shift effects keeping all markers
  
  # Compute G-matrix
  W_matrix <- Gmatrix(SNPmatrix = M.clean, method = "VanRaden") # , weights = W
  Ghat.blend <- G.tuneup(G = W_matrix, blend = TRUE, pblend = 0.02)$Gb
  Ginv.sparse <- G.inverse(G = Ghat.blend, sparse = TRUE)$Ginv
  
  # Loop for prediction
  for (j in 1:runs) {
    set.seed(j + as.integer(Sys.time()))
    tst <- sample(1:n, size = nt, replace = FALSE)
    yNA <- pheno$DF_BLUP
    yNA[tst] <- NA
    
    model <- asreml(fixed = DF_BLUP ~ 1, random = ~vm(Taxa, source = Ginv.sparse), data = pheno[-tst,])
    
    pred_gblup_asreml <- data.frame(
      y = pheno[tst, "DF_BLUP"], 
      predict(model, classify = "Taxa")$pvals[tst, c("Taxa", "predicted.value")]
    )
    
    corr.val[j, i] <- cor(as.numeric(pred_gblup_asreml$DF_BLUP), as.numeric(pred_gblup_asreml$predicted.value), use = "complete.obs")
  }
}

# Compute mean and standard errors
mean_corr <- colMeans(corr.val, na.rm = TRUE)
SE_corr <- apply(corr.val, 2, function(x) sd(x, na.rm = TRUE) / sqrt(runs))

# Create summary table
summary_table <- data.frame(
  Set = 1:10,
  mean_corr = mean_corr,
  SE_corr = SE_corr
)

# Print results
table_output <- capture.output(print(summary_table, row.names = FALSE))
writeLines(table_output)

View(summary_table)

mean(summary_table$mean_corr)

```

